File: Validate Binary Search Tree.cpp
// Leetcode Problem 98: Validate Binary Search Tree
// C++ CODE

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right)
 *         : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:

    bool validate(TreeNode* node, long long low, long long high) {
        if (node == nullptr) {
            return true;
        }

        // Node must be strictly inside the valid range
        if (node->val <= low || node->val >= high) {
            return false;
        }

        // Left subtree: values < node->val
        // Right subtree: values > node->val
        return validate(node->left, low, node->val) &&
               validate(node->right, node->val, high);
    }

    bool isValidBST(TreeNode* root) {
        return validate(root, LLONG_MIN, LLONG_MAX);
    }
};

File: Validate Binary Search Tree.java
// Leetcode Problem 98: Validate Binary Search Tree
// JAVA CODE

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    public boolean isValidBST(TreeNode root) {
        return validate(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    private boolean validate(TreeNode node, long low, long high) {
        if (node == null) {
            return true;
        }

        // Node value must be strictly inside the valid range
        if (node.val <= low || node.val >= high) {
            return false;
        }

        // Left subtree: values must be smaller than node.val
        // Right subtree: values must be greater than node.val
        return validate(node.left, low, node.val)
                && validate(node.right, node.val, high);
    }
}